plugins/vli-usbhub/fu-vli-usbhub-pd-device.c

```diff
@@ -0,0 +1,206 @@
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INVALID_FILE,
+			     "firmware too large, got 0x%x, expected <= 0x%x",
+			     (guint) g_bytes_get_size (fw),
+			     (guint) fu_device_get_firmware_size_max (device));
+		return NULL;
+	}
+
+	/* check is compatible with firmware */
+	fu_device_set_status (device, FWUPD_STATUS_DECOMPRESSING);
+	if (!fu_firmware_parse (firmware, fw, flags, error))
+		return NULL;
+	chip = fu_vli_usbhub_pd_firmware_get_chip (FU_VLI_USBHUB_PD_FIRMWARE (firmware));
+	if (self->chip != chip) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INVALID_FILE,
+			     "firmware incompatible, got %s, expected %s",
+			     fu_vli_usbhub_pd_chip_to_string (chip),
+			     fu_vli_usbhub_pd_chip_to_string (self->chip));
+		return NULL;
+	}
+
+	/* we could check this against flags */
+	g_debug ("parsed version: %s", fu_firmware_get_version (firmware));
+	return g_steal_pointer (&firmware);
+}
+
+static FuFirmware *
+fu_vli_usbhub_pd_device_read_firmware (FuDevice *device, GError **error)
+{
+	FuVliUsbhubDevice *parent = FU_VLI_USBHUB_DEVICE (fu_device_get_parent (device));
+	FuVliUsbhubPdDevice *self = FU_VLI_USBHUB_PD_DEVICE (device);
+	g_autoptr(GBytes) fw = NULL;
+	fu_device_set_status (FU_DEVICE (device), FWUPD_STATUS_DEVICE_VERIFY);
+	fw = fu_vli_usbhub_device_spi_read (parent,
+					    fu_vli_usbhub_pd_chip_get_offset (self->chip),
+					    fu_device_get_firmware_size_max (device),
+					    error);
+	if (fw == NULL)
+		return NULL;
+	return fu_firmware_new_from_bytes (fw);
+}
+
+static gboolean
+fu_vli_usbhub_pd_device_write_firmware (FuDevice *device,
+					FuFirmware *firmware,
+					FwupdInstallFlags flags,
+					GError **error)
+{
+	FuVliUsbhubPdDevice *self = FU_VLI_USBHUB_PD_DEVICE (device);
+	FuVliUsbhubDevice *parent = FU_VLI_USBHUB_DEVICE (fu_device_get_parent (device));
+	gsize bufsz = 0;
+	const guint8 *buf;
+	g_autoptr(GBytes) fw = NULL;
+
+	/* simple image */
+	fw = fu_firmware_get_image_default_bytes (firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* erase */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_ERASE);
+	buf = g_bytes_get_data (fw, &bufsz);
+	if (!fu_vli_usbhub_device_spi_erase (parent,
+					     fu_vli_usbhub_pd_chip_get_offset (self->chip),
+					     bufsz,
+					     error))
+		return FALSE;
+
+	/* write */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	if (!fu_vli_usbhub_device_spi_write (parent,
+					     fu_vli_usbhub_pd_chip_get_offset (self->chip),
+					     buf,
+					     bufsz,
+					     error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static void
+fu_vli_usbhub_pd_device_init (FuVliUsbhubPdDevice *self)
+{
+	fu_device_add_icon (FU_DEVICE (self), "audio-card");
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_CAN_VERIFY_IMAGE);
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_USABLE_DURING_UPDATE);
```

Comment:
well now that you're doing a composite cleanup that will cause the reset I don't think this is true anymore

Reply:
Ack, got it, thanks.